File: apps/mobile/ios/Fabric/Core/PairingURI.swift
import Foundation

/// Validation for a server address entered directly by the user.
enum GatewayBaseURL {
    static func parse(_ raw: String) -> URL? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard
            !trimmed.isEmpty,
            !trimmed.unicodeScalars.contains(where: { $0.value <= 32 }),
            let components = URLComponents(string: trimmed),
            components.scheme?.lowercased() == "http" || components.scheme?.lowercased() == "https",
            !(components.host ?? "").isEmpty,
            components.user == nil,
            components.password == nil,
            components.query == nil,
            components.fragment == nil
        else { return nil }
        return components.url
    }
}

enum PairingEnrollmentAuth: Equatable {
    case browser
    case local
}

/// A v2 QR's time-limited handoff. The raw value stays in memory only and is
/// never interpreted as a session token or a saved device credential.
struct PairingEnrollment: Equatable {
    let handle: String
    let auth: PairingEnrollmentAuth
}

/// Parsed `fabric://pair` payload from a pairing QR
/// (emitted by `fabric mobile`; contract in `fabric_cli/mobile_pairing.py`).
///
/// - `auth == "token"`: `token` is the session credential; connect directly.
/// - `auth == "gated"`: the gateway requires a provider login; the app asks
///   for username/password after the scan.
/// - version 2: an opaque, one-time `enrollment` handoff is recognized but
///   must not be misclassified as either of the legacy paths.
struct PairingPayload: Equatable {
    let baseURL: URL
    let gated: Bool
    let token: String?
    let enrollment: PairingEnrollment?

    private static let gatedKeys: Set<String> = ["v", "url", "auth"]
    private static let tokenKeys = gatedKeys.union(["token"])
    private static let enrollmentKeys: Set<String> = ["v", "url", "enrollment", "auth"]

    /// Parse either the canonical payload or the browser landing URL whose
    /// fragment contains that payload. Direct server addresses belong to
    /// `GatewayBaseURL`, not this machine-readable contract.
    static func parse(_ raw: String) -> PairingPayload? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let components = URLComponents(string: trimmed) else { return nil }
        if components.scheme?.lowercased() == "fabric" {
            return parsePayload(components)
        }

        guard
            components.scheme?.lowercased() == "http" || components.scheme?.lowercased() == "https",
            !(components.host ?? "").isEmpty,
            components.user == nil,
            components.password == nil,
            components.path == "/mobile/pair",
            components.query == nil,
            let fragment = components.percentEncodedFragment,
            let fragmentItems = parameters(from: fragment),
            Set(fragmentItems.keys) == ["pair"],
            let wrapped = fragmentItems["pair"],
            let payload = URLComponents(string: wrapped)
        else { return nil }
        return parsePayload(payload)
    }

    private static func parsePayload(_ components: URLComponents) -> PairingPayload? {
        guard
            components.scheme?.lowercased() == "fabric",
            components.host == "pair",
            components.user == nil,
            components.password == nil,
            components.port == nil,
            components.path.isEmpty,
            components.fragment == nil,
            let query = components.percentEncodedQuery,
            let params = parameters(from: query),
            let version = params["v"],
            let rawBaseURL = params["url"],
            let baseURL = GatewayBaseURL.parse(rawBaseURL)
        else { return nil }

        if version == "2" {
            guard
                Set(params.keys) == enrollmentKeys,
                baseURL.scheme?.lowercased() == "https",
                canonicalV2GatewayURL(rawBaseURL),
                let handle = params["enrollment"],
                validEnrollmentHandle(handle)
            else { return nil }
            let enrollmentAuth: PairingEnrollmentAuth
            switch params["auth"] {
            case "browser": enrollmentAuth = .browser
            case "local": enrollmentAuth = .local
            default: return nil
            }
            return PairingPayload(
                baseURL: baseURL,
                gated: false,
                token: nil,
                enrollment: PairingEnrollment(handle: handle, auth: enrollmentAuth)
            )
        }

        guard version == "1" else { return nil }

        switch params["auth"] {
        case "gated":
            guard Set(params.keys) == gatedKeys else { return nil }
            return PairingPayload(baseURL: baseURL, gated: true, token: nil, enrollment: nil)
        case "token":
            guard
                Set(params.keys) == tokenKeys,
                let token = params["token"],
                !token.isEmpty
            else { return nil }
            return PairingPayload(baseURL: baseURL, gated: false, token: token, enrollment: nil)
        default:
            return nil
        }
    }

    private static func validEnrollmentHandle(_ value: String) -> Bool {
        guard value.range(of: "^[A-Za-z0-9_-]{43,128}$", options: .regularExpression) != nil else {
            return false
        }
        return true
    }

    /// `URL` normalizes repeated trailing slashes. Preserve the machine-issued
    /// v2 wire grammar by checking `URLComponents` before that normalization.
    private static func canonicalV2GatewayURL(_ raw: String) -> Bool {
        guard
            raw == raw.trimmingCharacters(in: .whitespacesAndNewlines),
            !raw.unicodeScalars.contains(where: { $0.value <= 32 }),
            let components = URLComponents(string: raw),
            components.scheme?.lowercased() == "https",
            components.percentEncodedPath.isEmpty || components.percentEncodedPath == "/"
        else { return false }
        return true
    }

    private static func parameters(from encoded: String) -> [String: String]? {
        guard let items = URLComponents(string: "fabric://parameters?\(encoded)")?.queryItems else {
            return nil
        }
        var values: [String: String] = [:]
        for item in items {
            guard !item.name.isEmpty, let value = item.value, values[item.name] == nil else {
                return nil
            }
            values[item.name] = value
        }
        return values
    }
}

File: apps/mobile/android/app/src/main/kotlin/io/github/obliviousodin/fabric/mobile/core/PairingUri.kt
package io.github.obliviousodin.fabric.mobile.core

import java.net.URI
import java.net.URLDecoder

/** Validation for a server address entered directly by the user. */
object GatewayBaseUrl {
    fun parse(raw: String): String? {
        val trimmed = raw.trim().trimEnd('/')
        if (trimmed.isEmpty() || trimmed.any { it.code <= 32 }) return null
        val uri = runCatching { URI(trimmed) }.getOrNull() ?: return null
        val scheme = uri.scheme?.lowercase()
        if (
            scheme !in setOf("http", "https") ||
            uri.host.isNullOrEmpty() ||
            uri.userInfo != null ||
            uri.rawQuery != null ||
            uri.rawFragment != null
        ) return null
        // Cleartext is only allowed to a local/private host, matching iOS's
        // NSAllowsLocalNetworking. Android has no ATS equivalent and the
        // release build permits cleartext at the OS layer (see
        // res/xml/network_security_config.xml), so this is the app-layer guard
        // that keeps a plain-http socket from ever reaching a public gateway.
        if (scheme == "http" && !isLocalOrPrivateHost(uri.host)) return null
        return uri.toString().trimEnd('/')
    }

    private val ipv4 = Regex("""^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$""")

    /**
     * True when cleartext http to [rawHost] stays on the local machine or a
     * private/tailnet network: loopback, RFC1918, link-local, CGNAT
     * (100.64.0.0/10, Tailscale), IPv6 loopback/link-local/ULA, `.local`
     * mDNS names, and single-label hostnames. Public hosts return false and
     * must use https.
     */
    internal fun isLocalOrPrivateHost(rawHost: String?): Boolean {
        var host = rawHost?.trim()?.lowercase() ?: return false
        if (host.startsWith("[") && host.endsWith("]")) host = host.substring(1, host.length - 1)
        if (host.isEmpty()) return false
        if (host.contains(':')) return isPrivateIpv6(host)
        val match = ipv4.matchEntire(host)
        if (match != null) return isPrivateIpv4(match)
        if (!host.contains('.')) return true // single-label host (localhost, raspberrypi)
        return host.endsWith(".local")
    }

    private fun isPrivateIpv4(match: MatchResult): Boolean {
        val (a, b, c, d) = match.destructured
        val octets = listOf(a, b, c, d).map { it.toInt() }
        if (octets.any { it > 255 }) return false
        val (o0, o1) = octets
        return when {
            o0 == 127 -> true // loopback 127.0.0.0/8
            o0 == 10 -> true // private 10.0.0.0/8
            o0 == 172 && o1 in 16..31 -> true // private 172.16.0.0/12
            o0 == 192 && o1 == 168 -> true // private 192.168.0.0/16
            o0 == 169 && o1 == 254 -> true // link-local 169.254.0.0/16
            o0 == 100 && o1 in 64..127 -> true // CGNAT/Tailscale 100.64.0.0/10
            else -> false
        }
    }

    private fun isPrivateIpv6(host: String): Boolean = when {
        host == "::1" -> true // loopback
        Regex("^fe[89ab]").containsMatchIn(host) -> true // fe80::/10 link-local
        host.startsWith("fc") || host.startsWith("fd") -> true // fc00::/7 ULA
        else -> false
    }
}

/**
 * Parsed version-1 `fabric://pair` payload from a pairing QR
 * (emitted by `fabric mobile`; contract in fabric_cli/mobile_pairing.py).
 *
 * - `gated == false`: [token] is the session credential; connect directly.
 * - `gated == true`: the gateway requires a provider login; the app asks for
 *   username/password after the scan.
 */
data class PairingPayload(
    val baseUrl: String,
    val gated: Boolean,
    val token: String?,
) {
    companion object {
        private val gatedKeys = setOf("v", "url", "auth")
        private val tokenKeys = gatedKeys + "token"

        /**
         * Parse either the canonical payload or the browser landing URL whose
         * fragment contains that payload. Direct server addresses belong to
         * [GatewayBaseUrl], not this machine-readable contract.
         */
        fun parse(raw: String): PairingPayload? {
            val trimmed = raw.trim()
            val uri = runCatching { URI(trimmed) }.getOrNull() ?: return null
            if (uri.scheme?.lowercase() == "fabric") return parsePayload(uri)

            if (
                uri.scheme?.lowercase() !in setOf("http", "https") ||
                uri.host.isNullOrEmpty() ||
                uri.userInfo != null ||
                uri.path != "/mobile/pair" ||
                uri.rawQuery != null ||
                uri.rawFragment.isNullOrEmpty()
            ) return null
            val fragment = parseParameters(uri.rawFragment) ?: return null
            if (fragment.keys != setOf("pair")) return null
            val payloadUri = runCatching { URI(fragment.getValue("pair")) }.getOrNull()
                ?: return null
            return parsePayload(payloadUri)
        }

        private fun parsePayload(uri: URI): PairingPayload? {
            if (
                uri.scheme?.lowercase() != "fabric" ||
                uri.authority != "pair" ||
                uri.path.orEmpty().isNotEmpty() ||
                uri.rawFragment != null
            ) return null
            val params = parseParameters(uri.rawQuery) ?: return null
            if (params["v"] != "1") return null
            val baseUrl = params["url"]?.let(GatewayBaseUrl::parse) ?: return null

            return when (params["auth"]) {
                "gated" -> {
                    if (params.keys != gatedKeys) return null
                    PairingPayload(baseUrl = baseUrl, gated = true, token = null)
                }

                "token" -> {
                    val token = params["token"]?.takeIf { it.isNotEmpty() } ?: return null
                    if (params.keys != tokenKeys) return null
                    PairingPayload(baseUrl = baseUrl, gated = false, token = token)
                }

                else -> null
            }
        }

        private fun parseParameters(raw: String?): Map<String, String>? {
            if (raw.isNullOrEmpty()) return emptyMap()
            val values = mutableMapOf<String, String>()
            for (pair in raw.split('&')) {
                val index = pair.indexOf('=')
                if (index <= 0) return null
                val key = decode(pair.substring(0, index)) ?: return null
                val value = decode(pair.substring(index + 1)) ?: return null
                if (values.put(key, value) != null) return null
            }
            return values
        }

        private fun decode(value: String): String? = runCatching {
            URLDecoder.decode(value, "UTF-8")
        }.getOrNull()
    }
}
